# Exam 
## Exercice 2

In [103]:
import pandas as pd
import numpy as np
import myhelpers.preprocessing as prepro
import myhelpers.visualisation as viz

In [104]:
df_visits = pd.read_csv("visits.csv")
df_vehicles = pd.read_csv("vehicles.csv")

In [105]:
df_visits.head()

id_veh        date country    brand  \
0  74cafe94426a7c6ec045df5221f2be62  2019-03-05      FR  PEUGEOT   
1  84e38dbe5929b7282921293bf8e43e14  2018-04-10      FR  PEUGEOT   
2  18b782a6633885368542a3532f45201a  2017-10-18      FR  PEUGEOT   
3  d03fbb36360fa44aea10f8bb558e2150  2018-10-18      FR  PEUGEOT   
4  52c19bb214abdcf2fa86c9d163fc9adc  2017-10-24      FR  PEUGEOT   

                          dealer_id  mileage_km  client_amount  
0  79aa5331b115415d12fead30d19004ab     41460.0         125.53  
1  84e59a56e610f5febbd26fdfa83ea96d     13372.0           0.00  
2  9c84f3117d821977a8570f9f6c335fc5     12601.0           0.00  
3  050ad488d73415f8f74f255d9cd48ef9     26887.0           0.00  
4  21354d33d13f120a44d7a839c79ffc95     10684.0           0.00

In [106]:
df_vehicles.head()

id_veh    brand model segment  \
0  1cb0d81b6752d3589bf1e316e70d3716  PEUGEOT   308       C   
1  22007b293588cf266de0626eb84bb374  PEUGEOT   308       C   
2  a2f5f4a15024227d2b519296c1c38a13  PEUGEOT   208       B   
3  bcf528ff2dc2676f200827c22c3c4607  PEUGEOT   308       C   
4  a875f521d7fad145fccf433804f4af8f  PEUGEOT  2008       B   

     registration_date  
0  2017-03-24 00:00:00  
1  2017-11-14 00:00:00  
2  2018-06-01 00:00:00  
3  2017-12-01 00:00:00  
4  2016-03-22 00:00:00

In [107]:
#data = pd.merge(df_visits.drop('brand', axis=1), df_vehicles, left_on='id_veh', right_on='id_veh', how='left')

Combien de véhicules sont présents dans les deux datasets ? Combien de véhicules
distincts ?

In [108]:
df_vehicles['id_veh'].nunique()

184067

In [109]:
df_visits['id_veh'].nunique()

182178

In [110]:
n_veh = max(df_vehicles['id_veh'].nunique(), df_visits['id_veh'].nunique())
n_veh

184067

 Combien de véhicules ont été mis en circulation chaque année ?

In [111]:
df_vehicles['year'] = df_vehicles['registration_date'].apply(lambda x: x.split('-')[0])

In [112]:
df_vehicles.groupby('year').nunique()[['id_veh']]

id_veh
year        
2015   27788
2016   55166
2017   53397
2018   38049
2019    9669

 Combien de marques diérentes de véhicules contiennent les deux datasets ?

In [113]:
df_vehicles['brand'].nunique()

3

In [114]:
df_visits['brand'].nunique()

3

In [115]:
df_visits['brand'].unique()

array(['PEUGEOT', 'CITROEN', 'DS'], dtype=object)

Top 5 id_veh ayant eectué le plus de visites

In [116]:
df_visits.groupby('id_veh').count().sort_values(by='date',ascending=False).head()

date  country  brand  dealer_id  mileage_km  \
id_veh                                                                          
60854a8b74b0f914c57fb59ab74bca08    96       96     96         96          96   
3dd91066c952b3e8753fd9bf8aa52df3    96       96     96         96          93   
7c6dbe1c42b4467c13fdb69d7f19870c    74       74     74         74          74   
a4d9525abe19316d75625413278dcdd8    66       66     66         66          66   
76ce03fdfe4eb0ef3829ff770076e66a    43       43     43         43          43   

                                  client_amount  
id_veh                                           
60854a8b74b0f914c57fb59ab74bca08             96  
3dd91066c952b3e8753fd9bf8aa52df3             96  
7c6dbe1c42b4467c13fdb69d7f19870c             74  
a4d9525abe19316d75625413278dcdd8             66  
76ce03fdfe4eb0ef3829ff770076e66a             43

Top 5 dealer_id ayant reçu le plus de visites

In [117]:
df_visits.groupby('dealer_id').count().sort_values(by='date',ascending=False).head()

id_veh   date  country  brand  mileage_km  \
dealer_id                                                                     
8002ff83402f7062af83c861f90d94da   15843  15843    15843  15843       15705   
401d8f5bdcb33b19de1a5b67f3af60f1   13660  13660    13660  13660       13347   
d4895cafaf18aecb1b57b3a527a2d10c   11688  11688    11688  11688       11688   
a8f718ad8d09eb14f37130d1ee598568   11606  11606    11606  11606       11606   
78184dc764ce1b59a0260a6a05c6e415   10467  10467    10467  10467       10467   

                                  client_amount  
dealer_id                                        
8002ff83402f7062af83c861f90d94da          15843  
401d8f5bdcb33b19de1a5b67f3af60f1          13660  
d4895cafaf18aecb1b57b3a527a2d10c          11688  
a8f718ad8d09eb14f37130d1ee598568          11606  
78184dc764ce1b59a0260a6a05c6e415          10467

Quelle est la proportion de véhicules ayant eectué une visite lors des 18 derniers
mois (entre le 01/03/2018 et le 01/09/2019) 

In [118]:
df_visits['date'] = pd.to_datetime(df_visits['date'])

In [119]:
start_date = '01/03-2018'
end_date = '01/09/2019'
mask = (df_visits['date'] > start_date) & (df_visits['date'] <= end_date)

In [120]:
visite_18 = df_visits.loc[mask]['id_veh'].nunique()
visite_18

111587

In [121]:
visite_18/n_veh*100

60.623034003922484

Data preprocessing


In [122]:
data = pd.merge(df_visits.drop('brand', axis=1), df_vehicles, left_on='id_veh', right_on='id_veh', how='left')

In [123]:
data.head()

id_veh       date country  \
0  74cafe94426a7c6ec045df5221f2be62 2019-03-05      FR   
1  84e38dbe5929b7282921293bf8e43e14 2018-04-10      FR   
2  18b782a6633885368542a3532f45201a 2017-10-18      FR   
3  d03fbb36360fa44aea10f8bb558e2150 2018-10-18      FR   
4  52c19bb214abdcf2fa86c9d163fc9adc 2017-10-24      FR   

                          dealer_id  mileage_km  client_amount    brand model  \
0  79aa5331b115415d12fead30d19004ab     41460.0         125.53  PEUGEOT   308   
1  84e59a56e610f5febbd26fdfa83ea96d     13372.0           0.00  PEUGEOT  5008   
2  9c84f3117d821977a8570f9f6c335fc5     12601.0           0.00  PEUGEOT   208   
3  050ad488d73415f8f74f255d9cd48ef9     26887.0           0.00  PEUGEOT  2008   
4  21354d33d13f120a44d7a839c79ffc95     10684.0           0.00  PEUGEOT  2008   

  segment    registration_date  year  
0       C  2016-01-21 00:00:00  2016  
1       C  2017-06-19 00:00:00  2017  
2       B  2016-05-02 00:00:00  2016  
3       B  2016-02-22 00:00:00  2016  
4       B  2016-12-16 00:00:00  2016

In [124]:
data['registration_date'] = pd.to_datetime(data['registration_date'])

In [125]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 548606 entries, 0 to 548605
Data columns (total 11 columns):
id_veh               548606 non-null object
date                 548606 non-null datetime64[ns]
country              548606 non-null object
dealer_id            548606 non-null object
mileage_km           543698 non-null float64
client_amount        548606 non-null float64
brand                533937 non-null object
model                533823 non-null object
segment              531740 non-null object
registration_date    533937 non-null datetime64[ns]
year                 533937 non-null object
dtypes: datetime64[ns](2), float64(2), object(7)
memory usage: 35.6+ MB


In [126]:
data.describe()

mileage_km  client_amount
count  5.436980e+05  548606.000000
mean   7.659559e+04      73.506493
std    2.120807e+07     192.564559
min    0.000000e+00   -9055.682000
25%    7.287000e+03       0.000000
50%    1.818400e+04      26.338000
75%    3.476800e+04     110.405000
max    9.999667e+09    9760.762000

In [127]:
data.isnull().sum()

id_veh                   0
date                     0
country                  0
dealer_id                0
mileage_km            4908
client_amount            0
brand                14669
model                14783
segment              16866
registration_date    14669
year                 14669
dtype: int64

In [128]:
data = prepro.input_missing_values(data)

In [129]:
data.mileage_km = data['mileage_km'].fillna(value= data.mileage_km.median())

In [130]:
data.isnull().sum()

id_veh               0
date                 0
country              0
dealer_id            0
mileage_km           0
client_amount        0
brand                0
model                0
segment              0
registration_date    0
year                 0
dtype: int64

Feature engineering

In [131]:
today = pd.to_datetime("today")

In [132]:
data['age_veh'] = data.year.apply(lambda x: today.year - int(x))

In [133]:
data.head()

id_veh       date country  \
0  74cafe94426a7c6ec045df5221f2be62 2019-03-05      FR   
1  84e38dbe5929b7282921293bf8e43e14 2018-04-10      FR   
2  18b782a6633885368542a3532f45201a 2017-10-18      FR   
3  d03fbb36360fa44aea10f8bb558e2150 2018-10-18      FR   
4  52c19bb214abdcf2fa86c9d163fc9adc 2017-10-24      FR   

                          dealer_id  mileage_km  client_amount    brand model  \
0  79aa5331b115415d12fead30d19004ab     41460.0         125.53  PEUGEOT   308   
1  84e59a56e610f5febbd26fdfa83ea96d     13372.0           0.00  PEUGEOT  5008   
2  9c84f3117d821977a8570f9f6c335fc5     12601.0           0.00  PEUGEOT   208   
3  050ad488d73415f8f74f255d9cd48ef9     26887.0           0.00  PEUGEOT  2008   
4  21354d33d13f120a44d7a839c79ffc95     10684.0           0.00  PEUGEOT  2008   

  segment registration_date  year  age_veh  
0       C        2016-01-21  2016        4  
1       C        2017-06-19  2017        3  
2       B        2016-05-02  2016        4  
3       B        2016-02-22  2016        4  
4       B        2016-12-16  2016        4

Construction de la target

In [134]:
start_date

'01/03-2018'

In [135]:
end_date

'01/09/2019'

In [157]:
target = (~((data['date'] > start_date) & (data['date'] <= end_date))).astype(int)

In [158]:
X = data.copy()

In [162]:
X.head()

id_veh       date  \
0  74cafe94426a7c6ec045df5221f2be62 2019-03-05   
1  84e38dbe5929b7282921293bf8e43e14 2018-04-10   
2  18b782a6633885368542a3532f45201a 2017-10-18   
3  d03fbb36360fa44aea10f8bb558e2150 2018-10-18   
4  52c19bb214abdcf2fa86c9d163fc9adc 2017-10-24   

                          dealer_id  mileage_km  client_amount model  \
0  79aa5331b115415d12fead30d19004ab     41460.0         125.53   308   
1  84e59a56e610f5febbd26fdfa83ea96d     13372.0           0.00  5008   
2  9c84f3117d821977a8570f9f6c335fc5     12601.0           0.00   208   
3  050ad488d73415f8f74f255d9cd48ef9     26887.0           0.00  2008   
4  21354d33d13f120a44d7a839c79ffc95     10684.0           0.00  2008   

  registration_date  year  age_veh  segment_A  ...  segment_C  \
0        2016-01-21  2016        4          0  ...          1   
1        2017-06-19  2017        3          0  ...          1   
2        2016-05-02  2016        4          0  ...          0   
3        2016-02-22  2016        4          0  ...          0   
4        2016-12-16  2016        4          0  ...          0   

   segment_C-TRANSPORT  segment_D  segment_D-TRANSPORT  segment_E-TRANSPORT  \
0                    0          0                    0                    0   
1                    0          0                    0                    0   
2                    0          0                    0                    0   
3                    0          0                    0                    0   
4                    0          0                    0                    0   

   brand_CITROEN  brand_DS  brand_PEUGEOT  country_ES  country_FR  
0              0         0              1           0           1  
1              0         0              1           0           1  
2              0         0              1           0           1  
3              0         0              1           0           1  
4              0         0              1           0           1  

[5 rows x 21 columns]

Mettre en place un modèle de machine learning

In [163]:
import myhelpers.modelisation as hm

In [164]:
X = X[['mileage_km', 'year', 'age_veh']]

In [165]:
model = hm.rndForrestModel(X, target)

d:\python\python36-32\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


In [168]:
model['score_test']

0.6985332539402851

In [174]:
X = data.copy()
X = X[['mileage_km', 'year', 'age_veh', 'country', 'segment']]
X = prepro.addDummies(X, 'country')
X = prepro.addDummies(X, 'segment')
model = hm.rndForrestModel(X, target)
model['score_test']

d:\python\python36-32\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.7045788725377016

In [172]:
X = data.copy()
X = X[['mileage_km', 'year', 'client_amount', 'brand']]
X = prepro.addDummies(X, 'brand')
model = hm.rndForrestModel(X, target)
model['score_test']

d:\python\python36-32\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.7045849485362919

In [175]:
X = data.copy()
X = X[['mileage_km', 'year', 'age_veh', 'country']]
X = prepro.addDummies(X, 'country')
model = hm.rndForrestModel(X, target)
model['score_test']

d:\python\python36-32\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.6988674338627553

In [177]:
X = data.copy()
X = X[['mileage_km', 'year', 'age_veh', 'country', 'segment', 'client_amount', 'brand']]
X = prepro.addDummies(X, 'country')
X = prepro.addDummies(X, 'segment')
X = prepro.addDummies(X, 'brand')
model = hm.rndForrestModel(X, target)
model['score_test']

d:\python\python36-32\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.7081454837102478

In [178]:
X = data.copy()
X = X[['mileage_km', 'year', 'age_veh', 'country', 'segment', 'client_amount', 'brand']]
X = prepro.addDummies(X, 'country')
X = prepro.addDummies(X, 'segment')
X = prepro.addDummies(X, 'brand')
model = hm.get_model_RegLog(X, target)
model['score_test']

d:\python\python36-32\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
d:\python\python36-32\lib\site-packages\sklearn\svm\base.py:931: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
d:\python\python36-32\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


0.6233306193872963

In [179]:
X = data.copy()
X = X[['mileage_km', 'year', 'age_veh', 'segment', 'client_amount']]
X = prepro.addDummies(X, 'segment')
model = hm.get_model_RegLog(X, target)
model['score_test']

d:\python\python36-32\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
d:\python\python36-32\lib\site-packages\sklearn\svm\base.py:931: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
d:\python\python36-32\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


0.6233306193872963

----

Pour le feature engineerin on a ajouté l'age du vehicule
et des 'dummies' de quelques features avant de les faire passer au model

Le meilleur modele est le random forrest ['mileage_km', 'year', 'age_veh', 'country', 'segment', 'client_amount', 'brand'] avec un score de 70.81%

La metrique adopté est la precision parceque c'est un probleme de classification

Pour améliorer les résultats on peut changer les parametres du modele ou bien choisir d'autre modeles plus adaptés
On peut aussi faire un GridSearch pour avoir les meilleurs parametres

Pour mieux expliquer le churn on pourrait ajouter la qualité de service de la concession, si le vehicule est de premiere main ou non, etc

Cela depend mais on pourrait, l'integrer a une API en Django ou Flask, le mettre dans un contenaire(Docker) et le deployer dans le cloud(AWS, Azure, GC,...)